qiskit/circuit/quantumcircuitdata.py

```diff
@@ -59,52 +133,71 @@ def __setitem__(self, key, value):
         self._circuit._update_parameter_table(instruction)
 
     def insert(self, index, value):
-        self._circuit._data.insert(index, None)
-        self[index] = value
-
-    def __delitem__(self, i):
-        del self._circuit._data[i]
+        try:
+            # To consolidate validation to __setitem__, insert None and overwrite.
+            self._list.insert(index, None)
+            self[index] = value
+        except CircuitError as err:
+            del self._list[index]
+            raise err
 
-    def __len__(self):
-        return len(self._circuit._data)
+    def __delitem__(self, index):
+        del self._list[index]
```

Comment:
Could this be made the concrete definition in `VerifiedMutableSequence`?  I can't necessarily see a case where `del` would need to do any subclass-specific value validation, and the other defined subclass below uses the same method.

Reply:
Come to think of it, I _can_ think of a case - If `Qregs` manages the registers, and a register is deleted from the list, it should presumably call back to the circuit to delete references to the contained `Bit` instances, and raise a noisy exception if any current instructions operate on those bits.

Reply:
Good catch, I think there may be a case that I'm missing here (maybe removing a register that has old-style bits needs to be verboten until old-style bits are removed, or removing a register that's used in a conditional of a subsequent instruction) but if not, this could be made concrete.